browser/upgrade_detector/upgrade_when_idle.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/upgrade_detector/upgrade_when_idle.h"
+
+#include "base/command_line.h"
+#include "base/time/time.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/first_run/upgrade_util.h"
+#include "chrome/browser/lifetime/application_lifetime.h"
+#include "chrome/browser/profiles/profile_manager.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "chrome/browser/upgrade_detector/upgrade_detector.h"
+#include "chrome/common/chrome_switches.h"
+#include "components/browsing_data/core/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "ui/base/idle/idle.h"
+
+namespace {
+
+// How much idle time (since last input event) must have passed until we
+// restart the browser when an update is available and no window is open.
+constexpr int kIdleAmount = 3;  // Minutes (or seconds, if testing).
+
+bool IsTesting() {
+  // Implementation similar to IsTesting() in upgrade_detector_impl.cc.
+  const base::CommandLine& cmd_line = *base::CommandLine::ForCurrentProcess();
+  return cmd_line.HasSwitch(switches::kSimulateUpgrade) ||
+         cmd_line.HasSwitch(switches::kCheckForUpdateIntervalSec) ||
+         cmd_line.HasSwitch(switches::kSimulateCriticalUpdate) ||
+         cmd_line.HasSwitch(switches::kSimulateOutdated) ||
+         cmd_line.HasSwitch(switches::kSimulateOutdatedNoAU);
+}
+
+bool AreAnyBrowsersOpen() {
+  return BrowserList::GetInstance()->size() > 0;
+}
+
+bool AreAnyClearDataOnExitSettingsEnabled() {
+  ProfileManager* profile_manager = g_browser_process->profile_manager();
+  if (!profile_manager) {
+    // This can happen during tests.
+    return false;
+  }
+  for (Profile* profile : profile_manager->GetLoadedProfiles()) {
+    PrefService* prefs = profile->GetPrefs();
+    // Check if any of the "clear on exit" preferences are enabled
+    if (prefs->GetBoolean(browsing_data::prefs::kDeleteBrowsingHistoryOnExit) ||
```

Comment:
this seems fragile, if any new settings are added they probably won't make it here. At the very least there should be a `LINT.IfChange` in components/browsing_data/core/pref_names.h (separating out `kDeleteBraveLeoHistory`), but maybe a better option would be something like `BraveClearBrowsingData::WillClearOnExit`?